src/main/java/com/ecore/roles/service/implementation/MembershipsServiceImplementation.java

```diff
@@ -20,36 +23,51 @@
 
 @Log4j2
 @Service
-public class MembershipsServiceImpl implements MembershipsService {
+public class MembershipsServiceImplementation implements MembershipsService {
 
     private final MembershipRepository membershipRepository;
     private final RoleRepository roleRepository;
+    private final TeamsClient teamsClient;
 
     @Autowired
-    public MembershipsServiceImpl(
+    public MembershipsServiceImplementation(
+            TeamsClient teamsClient,
             MembershipRepository membershipRepository,
             RoleRepository roleRepository) {
         this.membershipRepository = membershipRepository;
         this.roleRepository = roleRepository;
+        this.teamsClient = teamsClient;
     }
 
     @Override
-    public Membership assignRoleToMembership(@NonNull Membership m) {
-
-        UUID roleId = ofNullable(m.getRole()).map(Role::getId)
+    public Membership assignRoleToMembership(@NonNull Membership membership) {
+        UUID roleId = ofNullable(membership.getRole()).map(Role::getId)
                 .orElseThrow(() -> new InvalidArgumentException(Role.class));
+        UUID teamId = membership.getTeamId();
+        UUID userId = membership.getUserId();
 
-        if (membershipRepository.findByUserIdAndTeamId(m.getUserId(), m.getTeamId())
+        if (membershipRepository.findByUserIdAndTeamId(membership.getUserId(), teamId)
                 .isPresent()) {
             throw new ResourceExistsException(Membership.class);
         }
 
         roleRepository.findById(roleId).orElseThrow(() -> new ResourceNotFoundException(Role.class, roleId));
-        return membershipRepository.save(m);
+
+        if (teamsClient.getTeam(teamId).getBody() == null) {
+            throw new ResourceNotFoundException(Team.class, teamId);
+        }
+
+        if (!teamsClient.getTeam(teamId).getBody().getTeamMemberIds().contains(userId)) {
+            throw new InvalidObjectException(Membership.class,
+                    "The provided user doesn't belong to the provided team");
+        }
```

Comment:
The function `teamsClient.getTeam(teamId).getBody().` was called multiple times in the same function. It would be possible create a separate variable and use it, instead off calling this function multiple times. 

```suggestion
        Team team = teamsClient.getTeam(teamId).getBody();
        
        if (team == null) {
            throw new ResourceNotFoundException(Team.class, teamId);
        }

        if (!team.getTeamMemberIds().contains(userId)) {
            throw new InvalidObjectException(Membership.class,
                    "The provided user doesn't belong to the provided team");
        }
```

It may lead to a NullPointerException on the automated testing, but it worths the shot anyway